filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,285 @@
+					}
+				}
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "reading kafka input config")
+	}
+
+	saramaConfig, err := newSaramaConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sarama config")
+	}
+	consumerGroup, err :=
+		sarama.NewConsumerGroup(config.Hosts, config.GroupID, saramaConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing kafka consumer group")
+	}
+
+	input := &kafkaInput{
+		config:        config,
+		rawConfig:     cfg,
+		context:       inputContext,
+		outlet:        out,
+		consumerGroup: consumerGroup,
+		sessionState:  sessionState,
+		log:           logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+// A helper to safely use the current sarama session for the duration of the
+// given callback. Used when ACKing messages outside the body of the main
+// sarama callbacks. The session parameter may be nil if there is no active
+// session.
+func (state *kafkaSessionState) accessSession(
+	fn func(session sarama.ConsumerGroupSession),
+) {
+	state.mutex.Lock()
+	state.waitGroup.Add(1)
+	session := state.session
+	state.mutex.Unlock()
+	defer state.waitGroup.Done()
+	fn(session)
+}
+
+// A helper to safely set the session field after waiting on any pending
+// operations.
+func (state *kafkaSessionState) setSession(sess sarama.ConsumerGroupSession) {
+	state.mutex.Lock()
+	// Once we claim the mutex we still wait for any pending ACKs to be
+	// sent. (These may well fail if the session is ending, but that's better
+	// than calling a stale pointer.)
+	state.waitGroup.Wait()
+	state.session = sess
+	state.mutex.Unlock()
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		// Track errors
+		go func() {
+			for err := range input.consumerGroup.Errors() {
+				input.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: input}
+
+				// Sarama uses standard go contexts to control cancellation, so we need
+				// to wrap our input context channel in that interface.
+				err := input.consumerGroup.Consume(
+					doneChannelContext(input.context.Done), input.config.Topics, handler)
+				if err != nil {
+					input.log.Errorw("Kafka consume error", "error", err)
+				}
+
+				// If Consume returned because the context was cancelled, don't resume.
+				select {
+				case <-input.context.Done:
+					return
+				default:
+				}
+			}
+		}()
+	})
+}
+
+// Wait shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Wait() {
+	input.Stop()
+	// Wait for the consumer group to shut down
+	input.consumerGroup.Close()
+}
+
+// Stop shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Stop() {
+	close(input.context.Done)
```

Comment:
I think `filebeat/input.Runner` already closes the channel on stop. If so, we have a chance of creating a panic here.

I think long-term 'Stop' becomes obsolete, if we make more use of context (or similar) to signal shutdown to run loops.

Reply:
It took a while to convince myself that they're the same channel and that the program flow always closes it, but I think you're right. I agree the `Stop` mechanism is kind of begging for a refactor right now and would probably be better off clarified and / or removed (since it's hard to follow who's responsible for closing which resource).

For now I've corrected this to closing `input.outlet`, and added a comment clarifying why explicitly closing the kafka consumer group is redundant here.